SofaKernel/modules/Sofa.Geometry/src/sofa/geometry/Triangle.h

```diff
@@ -90,8 +90,110 @@ struct Triangle
         constexpr Node n{};
         static_assert(std::distance(std::begin(n), std::end(n)) == 3, "Triangle normal can only be computed in 3 dimensions.");
 
-        return (n1 - n0).cross(n2 - n0);
+        // Vec gives access to cross() and operator-
+        if constexpr (std::is_same_v < Node, sofa::type::Vec<3, T> >)
+        {
+            const auto a = n1 - n0;
+            const auto b = n2 - n0;
+
+            return a.cross(b);
+        }
+        else
+        {
+            Node a{}, b{};
+            std::transform(n1.cbegin(), n1.cend(), n0.cbegin(), a.begin(), std::minus<T>());
+            std::transform(n2.cbegin(), n2.cend(), n0.cbegin(), b.begin(), std::minus<T>());
+
+            return Node{ a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2] , a[0] * b[1] - a[1] * b[0] };
+        }
+
     }
+
+    /**
+    * @brief	Test if a ray intersects a triangle, and gives barycentric coordinates of the intersection if applicable
+    * @remark   Implementation for 3D only
+    * @tparam   Node iterable container (or sofa::type::Vec with cross() and norm())
+    * @tparam   T scalar
+    * @param	n0,n1,n2 nodes of the triangle
+    * @param	t, u, v barycentric coefficients of the potential intersection in the triangle
+    * @return	either if the given ray intersects the given triangle or not
+    */
+    template<typename TReal>
+    [[nodiscard]]
+    static constexpr bool rayIntersection(const sofa::type::Vec<3, TReal>& n0, const sofa::type::Vec<3, TReal>& n1, const sofa::type::Vec<3, TReal>& n2, const sofa::type::Vec<3, TReal>& origin, const sofa::type::Vec<3, TReal>& direction, TReal& t, TReal& u, TReal& v)
+    {
+
+        static constexpr TReal epsilon = std::numeric_limits<TReal>::epsilon();
+        static constexpr TReal zero = static_cast<TReal>(0);
+        static constexpr TReal one = static_cast<TReal>(1);
+
+        t = 0; u = 0; v = 0;
+
+        const auto e0 = n1 - n0;
+        const auto e1 = n2 - n0;
+
+        sofa::type::Vector3 tvec(type::NOINIT);
+        sofa::type::Vector3 pvec(type::NOINIT);
+        sofa::type::Vector3 qvec(type::NOINIT);
+        TReal det;
+        TReal inv_det;
+
+        pvec = sofa::type::cross(direction, e1);
+
+        det = sofa::type::dot(e0, pvec);
+        if constexpr(std::is_floating_point_v<TReal>)
```

Comment:
I did a mistake on the isnan (shoud be on the inv_det) .🤪
but I want to try a little optimisation (but surely useless?) that if TReal is a floating point then if the det is 0, the result of the division is NaN (whereas if Treal is an integer it will throw an exception).
It would save one test (or a call to fabs()) in the end (but maybe it is a pointless optimization...... 👺)